keps/498-GroupScheduling/README.md

```diff
@@ -0,0 +1,395 @@
+individually at any time.
+
+### Example JobSet spec using group scheduling for all replicated jobs:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: my-jobset
+  namespace: default
+spec:
+  # main containers in pods part of replicatedJob `workers` will be blocked
+  # from execution until all pods have started, or timeout after 5min.
+  groupSchedulingConfig:
+    timeoutAfterSeconds: 300
+    targetReplicatedJobs:
+    - workers
+  replicatedJobs:
+    - name: workers
+      replicas: 3
+      template:
+        spec:
+          backoffLimit: 0
+          completions: 100
+          parallelism: 100
+          template:
+            spec:
+              containers:
+                - name: pytorch
+                  image: pytorch/pytorch:latest
+                  command:
+                    - python3
+                    - train.py
+    - name: auxiliary
+      replicas: 1
+      template:
+        spec:
+          backoffLimit: 0
+          completions: 1
+          parallelism: 1
+          template:
+            spec:
+              containers:
+                - name: auxiliary
+                  image: python:3.10
+                  command:
+                    - python3
+                    - run.py
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### Proposed Group Scheduling API
+
+```go
+
+// JobSetSpec defines the desired state of JobSet
+type JobSetSpec struct {
+        …
+        // GroupSchedulingConfig defines the desired group-scheduling behavior
+        // for the JobSet. If unspecified, every pod part of this JobSet will
+        // start as soon as possible, without waiting for the others.
+        GroupSchedulingConfig *GroupSchedulingConfig `json:groupSchedulingConfig`
+}
+
+// GroupSchedulingConfig defines the desired group-scheduling behavior for the
+// JobSet. If defined, the main container in pods part of the
+// TargetReplicatedJobs will be blocked from executing until all pods in these
+// jobs have started.
+type GroupSchedulingConfig struct {
+        // Timeout defines the period after which the injected initContainer
+        // (which blocks execution of the main container until all pods are started)
+        // will timeout and exit with an error if not all pods have started yet.
+        TimeoutAfterSeconds *int32 `json:"timeoutAfterSeconds"
```

Comment:
Reading 5 minutes to mean 300 s I wonder if we should consider durations as an API?

I guess k8s has a lot of apis in seconds so maybe its not a big deal.